src/Layer.js

```diff
@@ -43,56 +48,78 @@ export default class Layer {
         this._checkId(id);
         this._checkSource(source);
         this._checkViz(viz);
-        this._init(id, source, viz);
-    }
-
-    _init (id, source, viz) {
-        viz._boundLayer = this;
-        this._context = new Promise((resolve) => {
-            this._contextInitialize = resolve;
-        });
 
-        /* Custom Layer API attributes:
+        /* Mapbox GL Custom Layer API attributes:
           - id: string
           - type: "custom"
         */
         this.id = id;
         this.type = 'custom';
 
-        this.metadata = null;
-        this._state = 'init';
+        viz._boundLayer = this;
+        this._context = new Promise(resolve => {
+            this._contextInitialize = resolve;
+        });
+
+        this._state = states.INIT;
         this._visible = true;
-        this._isLoaded = false;
-        this._matrix = null;
-        this._fireUpdateOnNextRender = false;
         this._emitter = mitt();
-        this._oldDataframes = new Set();
         this._renderLayer = new RenderLayer();
-        this._atomicChangeUID = 0;
 
-        this.update(source, viz);
+        // Use an UID to detect that a new call to `Layer.update()` is overriding an old (uncommitted) one
+        this._majorNextUID = 0;
+        this._majorCurrentUID = null;
+        this._minorNextUID = 0;
+        this._minorCurrentUID = null;
+
+        this._sourcePromise = this.update(source, viz);
     }
 
     /**
-     * Get layer visibility. Can be 'visible' or 'none'.
-     * @readonly
+     * Layer visibility property.
+     *
+     * @type {boolean}
+     * @memberof carto.Layer
+     * @instance
+     * @api
      */
-    get visibility () {
-        return this._visible ? layerVisibility.VISIBLE : layerVisibility.HIDDEN;
+    get visible () {
+        return this._visible;
+    }
+
+    set visible (visible) {
+        this.map.setLayoutProperty(this.id, 'visibility', visible ? 'visible' : 'none');
+        this._visible = visible;
     }
 
     /**
-     * Get layer visibility. Can be true or false.
-     * @readonly
+     * Change layer visibility to visible
+     *
```

Comment:
Maybe we could add a small note about what's a mayor and a minor change / UID

Reply:
:+1: